tensorflow/lite/micro/kernels/gather.cc

```diff
@@ -0,0 +1,212 @@
+  }
+  return context->ResizeTensor(context, output, output_shape);
+}
+
+template <typename InputT, typename PositionsT>
+TfLiteStatus Gather(const TfLiteGatherParams& params, const TfLiteTensor* input,
+                    const TfLiteTensor* positions, TfLiteTensor* output) {
+  tflite::GatherParams op_params;
+  op_params.axis = params.axis;
+  optimized_ops::Gather(op_params, GetTensorShape(input),
+                        GetTensorData<InputT>(input), GetTensorShape(positions),
+                        GetTensorData<PositionsT>(positions),
+                        GetTensorShape(output), GetTensorData<InputT>(output));
+  return kTfLiteOk;
+}
+
+template <typename PositionT>
+TfLiteStatus GatherStrings(TfLiteContext* context, const TfLiteTensor* input,
+                           const TfLiteTensor* positions,
+                           TfLiteTensor* output) {
+  DynamicBuffer buffer;
+  const PositionT* indexes = GetTensorData<PositionT>(positions);
+  const PositionT num_strings = GetStringCount(input);
+  const int num_indexes = NumElements(positions);
+
+  for (int i = 0; i < num_indexes; ++i) {
+    const PositionT pos = indexes[i];
+    TF_LITE_ENSURE(context, pos < num_strings);
+    const auto string_ref = GetString(input, pos);
+    buffer.AddString(string_ref.str, string_ref.len);
+  }
+  buffer.WriteToTensor(output, /*new_shape=*/nullptr);
+  return kTfLiteOk;
+}
+
+TfLiteStatus Eval(TfLiteContext* context, TfLiteNode* node) {
+  const auto* params =
+      reinterpret_cast<const TfLiteGatherParams*>(node->builtin_data);
+  const TfLiteTensor* input;
+  TF_LITE_ENSURE_OK(context, GetInputSafe(context, node, kInputTensor, &input));
+  const TfLiteTensor* positions;
+  TF_LITE_ENSURE_OK(context,
+                    GetInputSafe(context, node, kInputPositions, &positions));
+  TfLiteTensor* output;
+  TF_LITE_ENSURE_OK(context,
+                    GetOutputSafe(context, node, kOutputTensor, &output));
+
+  if (positions->type == kTfLiteInt32) {
+    switch (input->type) {
+      case kTfLiteFloat32:
+        return Gather<float, int32_t>(*params, input, positions, output);
+      case kTfLiteUInt8:
+        return Gather<uint8_t, int32_t>(*params, input, positions, output);
+      case kTfLiteInt8:
+        return Gather<int8_t, int32_t>(*params, input, positions, output);
+      case kTfLiteInt16:
+        return Gather<int16_t, int32_t>(*params, input, positions, output);
+      case kTfLiteInt32:
+        return Gather<int32_t, int32_t>(*params, input, positions, output);
+      case kTfLiteInt64:
+        return Gather<int64_t, int32_t>(*params, input, positions, output);
+      case kTfLiteBool:
+        return Gather<bool, int32_t>(*params, input, positions, output);
+      case kTfLiteString:
+        return GatherStrings<int32_t>(context, input, positions, output);
+      default:
+        context->ReportError(context, "Type '%s' is not supported by gather.",
```

Comment:
We'll need to replace this with the error report macro instead, to avoid triggering the presubmit check.

Reply:
Now the code uses the TF_LITE_KERNEL_LOG macro.